### DISCUSSION TITLE
What's the differences between jspsych eyetracking  plugins and xiaozhi yang, krajbich eye tracking plugins?

### DISCUSSION THREAD

**Discussion by zaizibai** [jspsych/jsPsych#1892] [Q&A]
I want to perform an online eye-tracking study, and I noticed currently there are two tools that can help users to use jspsych to perform online eye-tracking study. One is the built-in jspsych eyetracking plugin and the other is the tool developed by xiaozhi yang and ian krajbich. What's the differences between those two tools? Thank you very much.

**Comment by myrtevos** [comment]
I'd like to second this question. I'm writing up some recent work I've done with eye-tracking in jsPsych, and I think the forked, modified version of webgazer used in v6.3.1. gave us markedly better results - but it's hard to figure out why/how :)

Having looked at the changes you made to webgazer in early April, and the discussion in issue #1700, it's not yet entirely clear to me what's going on. And like zaizibai said, Xiaozhi Yang published a preprint in December in which they mention modifying `loop()` to improve temporal resolution, but they're not explicit about what they did exactly, and I can't find the change in their code. I'd be very interested to know whether you both solved the same problem, using similar or different approaches.

Any insights you can give here are much appreciated. Thanks!

**Comment by jodeleeuw** [comment]
I'm not sure if I can *fully* explain the differences between my implementation in jsPsych 6.3.1 and the Yang & Krajbich implementation. I've corresponded with Xiaozhi about the work, but we haven't done a deep dive into all the differences. I can give a high-level overview.

For me, the major difference is just the API. We introduced [extensions](https://www.jspsych.org/extensions/extensions/) in jsPsych 6.3.0 in order to support a mode where eye tracking could be layered on top of any existing plugin. Yang & Krajbich wrote plugins specific to eye tracking. There are (dis)advantages to each approach. Custom eye tracking plugins give you more control over exactly when eye tracking is initialized and how the data are collected. The extensions approach makes it easier to layer eye tracking on top of all the existing functionality of jsPsych. (But even with extensions we still needed custom plugins for things like calibration, validation, and webcam initialization.)

As far as performance goes, I think we both applied similar modifications to webgazer. Xiaozhi was very gracious in sharing ideas as I was trying to make improvements from 6.3.0 -> 6.3.1, and we arrived at similar conclusions. Webgazer relies on the `requestAnimationFrame` cycle to update the processing on the video data. For reasons that I only partially understand, this can create a real bottleneck in the code and actually slow down the sampling rate. The major change we both made to our respective forks is that we disabled webgazer's automatic loop so that webgazer is no longer trying to provide an updated prediction with every animation frame, and instead we just invoke webgazer's prediction algorithm at a regular interval. This seems to actually speed up the rate at which calculations can be done. And, perhaps even more importantly, using `requestAnimationFrame` was causing blocking in jsPsych's timing, so if a participant had a particularly poor computer -- or even a good one -- the timing of experiments could become really bad really quickly (see #1700). I think our forks implement this change in slightly different ways.

**Reply by jodeleeuw** [reply]
Oh, and one addendum here:

In the future it looks like browsers may adopt [a new processing method](https://web.dev/requestvideoframecallback-rvfc/) that would allow a function to be invoked on every *video* frame instead of every *animation* frame. Once this happens, we can modify webgazer to use this and I think performance will be bumped up again.

Current adoption status: https://caniuse.com/mdn-api_htmlvideoelement_requestvideoframecallback

**Reply by zaizibai** [reply]
Thank you for making such a convenient tool! Really appreciate it. I have two questions about the data recorded by webgazer. The first one is where is the base point of the coordinate system? Is it located in the left-down corner of the screen or somewhere else?   The second one is I found some data's time is a negative value. is it the gaze data that before the stimulus-onset? Thank you very much.

**Comment by myrtevos** [comment]
Thank you so much once again for your prompt and excellent reply! (Thanks also to both you and Xiaozhi for wading in and figuring this out.) This is exactly what I wanted to know. Excited to hear that there's prospects for further performance improvement - I'm planning a lot more work with eye tracking in jsPsych over the next year :)
